---
title: "JoelleFitzgerald_Team1_Step2"
author: "Joelle Fitzgerald"
date: "11/2/2022"
output: html_document
---

## HIDS 503 Final Project Part 2: Enrichment Analysis 

## Goal 

To perform Enrichment Analysis in R with the help of the R package EnrichR.

```{r setup, include=FALSE}

# Installing appropriate packages
knitr::opts_chunk$set(echo = TRUE)
install.packages("enrichR")
install.packages("openxlsx")# independant of java
install.packages("devtools") 
devtools::install_github("wjawaid/enrichR")
library(enrichR)
library(openxlsx)

```


# Sub-set top differentially expressed genes

Cleared workspace & read in the T-Test results file.
```{r}
#Read in the T-Test results file
ttestResults <- read.csv(file = "output/JoelleFitzgerald_Team1_Step1_Ttest.csv")

#check to make sure p-value column is imported as numeric 
#sort by p-value (just in case the results are not sorted by p-value)
ttestResultsSorted <- dplyr::arrange(ttestResults, Pvalue)

#find rows with FDR < 0.01
whichSig <- which(ttestResultsSorted$FDR <= 0.01) 

#Short list sig results
ttestResultsSig <- ttestResultsSorted[whichSig, ] 

### Export short listed results
write.table(x = ttestResultsSig, 
            file = "output/Joelle_Fitzgerald_Step1_Ttest_Shortlisted.tsv", 
            quote = F, sep = "\t")

##### First column is a list of features in this format : ID|PrimaryBladderCancerType. 
#### Use string split strsplit() function to extract gene names
funcSplit <- function(featureX) {
  f1 <- unlist(strsplit(x = featureX, split = "|", fixed = TRUE))
  f2 <- f1[2]
  return(f2)
}

# Use apply() function to run the split on every row, its faster version of a loop
Status1 <- apply(X = as.matrix(ttestResultsSig$Feature), 
                    MARGIN = 1, FUN = funcSplit)

head(Status1) # Visual of split function results for disease status & corresponding genes

#print length of short listed gene names,  results
length(Status1)

### Export list of top 20 gene names
write.table(x = head(Status1,20), 
            file = "output/Joelle_Fitzgerald_Step1_SigDiff.tsv", 
            quote = F, sep = "\t")

 
```

## Given

A set of T-Test results from a Gene expression group comparison analysis. This analysis was done on comparison of pre-cancerous tissue (“Bladder Cancer Surrounding Mucosa”) vs Normal tissue (baseline group) samples.

## Import T-test results so that we can short list further
```{r}
#folder that contains group comparison results
BladderCancer_tTest <- "output/Joelle_Fitzgerald_Step1_Ttest_Shortlisted.tsv"

# read in group comparison results
result1 <- read.csv(file = BladderCancer_tTest, row.names = 1, sep="\t", header = TRUE , stringsAsFactors = F)
```

# Filtering/shortlisting based on p-value generated by t-Test 

Short list results based in p-value cut off (pvalue <= 0.01)

```{r , eval=FALSE}
pValueCutOff <- 0.01
which1 <- (as.numeric(p.adjust(result1$Pvalue, method='BH')) <= pValueCutOff)
table(which1) #273 genes TRUE

resultShort <- result1[which1, ] #short listed genes 

```


## Clean gene names

```{r}
pValueCutOff <- 0.01
which1 <- (as.numeric(p.adjust(result1$Pvalue, method='BH')) <= pValueCutOff)
table(which1) #273 genes TRUE
resultShort <- result1[which1, ] #short listed genes 

funcSplit <- function(rep_gene) {
    rep_gene_split <- unlist(strsplit(x = rep_gene, 
                                      split = "|", 
                                      fixed = TRUE))
    gene <- rep_gene_split[2]
    return(gene)
}
geneListSplit <- apply(X = as.matrix((resultShort$Feature)), 
                       MARGIN = 1, FUN = funcSplit ) # 250 genes
head(geneListSplit) #cleaned gene names

#remove duplicates
geneListSplit1 <- unique(geneListSplit) # 250 genes 

# remove NA value
geneListFinal <- na.omit(geneListSplit1) # 250 genes

#print number of unique genes
length(geneListFinal)
write.table(x = geneListFinal, 
          file = "output/JoelleFitzgerald_Team1_Step2_Shortlisted_DEGs2.tsv",
          quote = F, sep = "\t", row.names = FALSE, col.names = F) # 250 unique genes


```

## Export the short listed results for reference

```{r}
#length of short listed results, 273 genes 
nrow(resultShort)

write.csv(x = resultShort, file = "output/JoelleFitzgerald_TTest_Shortlist2.csv")

```

## Load Databases for Enrichr R package and testing connection

```{r}
#checking if EnrichR website and packing are working
#dbs <- enrichR::listEnrichrDbs()  #total number of databases available = 200+ 

#testing if EnrichR package is working
#testOutput <- enrichR::enrichr(genes = c("Runx1", "Gfi1", "Gfi1b", "Spi1", "Gata1", "Kdr"), databases = "KEGG_2021_Human")
#head(testOutput[[1]])

#List of databases for which enrichment analysis will be run
dblist <- read.csv(file = "input/2022-EnrichR-Databases.txt", 
                    header = F, stringsAsFactors = F)
#34 databses
dim(dblist) 

data.frame(dblist)

head(dblist1)
```

## Calling the function to run Enrichment


```{r}

# set output file name
outputFile1 <- paste("output/JoelleFitzgerald_Team1_Step2", "_EnrichR2.xlsx", sep="")

#Load R script into the environment
source(file = "input/functionEnrichment.R") 

#call function to run Enrichment
functionEnrichment(dblist1, geneListFinal, outputFile1)

```
